mysql/toddl.go

```diff
@@ -191,3 +192,35 @@ func cvtPrimaryKeys(conv *internal.Conv, srcTable string, srcKeys []schema.Key)
 	}
 	return spKeys
 }
+
+func cvtForeignKeys(conv *internal.Conv, srcTable string, srcKeys []schema.ForeignKey) []ddl.Foreignkey {
+	var spKeys []ddl.Foreignkey
+	for _, key := range srcKeys {
+		if len(key.Columns) != len(key.ReferColumns) {
+			conv.Unexpected(fmt.Sprintf("ConvertForeignKeys: columns and referColumns don't have the same lengths: len(columns)=%d, len(referColumns)=%d for source table: %s, referenced table: %s", len(key.Columns), len(key.ReferColumns), srcTable, key.ReferTable))
+			continue
+		}
+		spReferTable, err := internal.GetSpannerTable(conv, key.ReferTable)
+		if err != nil {
+			conv.Unexpected(fmt.Sprintf("Can't map foreign key for source table: %s, referenced table: %s", srcTable, key.ReferTable))
+			continue
+		}
+		var spCols, spReferCols []string
+		for i, col := range key.Columns {
+			spCol, err1 := internal.GetSpannerCol(conv, srcTable, col, false)
+			spReferCol, err2 := internal.GetSpannerCol(conv, key.ReferTable, key.ReferColumns[i], false)
+			if err1 != nil || err2 != nil {
+				conv.Unexpected(fmt.Sprintf("Can't map foreign key for table: %s, referenced table: %s, column: %s", srcTable, key.ReferTable, col))
+				continue
+			}
+			spCols = append(spCols, spCol)
+			spReferCols = append(spReferCols, spReferCol)
+		}
+		spKey := ddl.Foreignkey{Columns: spCols,
+			Name:         key.Name,
```

Comment:
Is key.Name always a valid Spanner foreign key constraint name?
Should we map these like we do with table names and col names i.e. using an equivalent to GetSpannerCol?

Reply:
Your concern is correct, key.Name is not always valid,

i tried some example with key name starting with number, which is valid for mysql but spanner throws error,

I will update the Name with GetSpannerKeyName. 

One challenge will be in collision detection to iterate over all the keys, as it is just an array and not a map.

Reply:
Hey Nevin,

I have added an implementation for GetSpannerKeyName:

here is the snippet:
```
func GetSpannerKeyName(conv *Conv, srcKeyName string, spKeys []ddl.Foreignkey) (string, error) {
	if srcKeyName == "" {
		return "", nil
	}
	spKeyName, _ := FixName(srcKeyName)

	// check wether the spKeyName is used before
	found := false
	for _, spKey := range spKeys {
		if spKey.Name == spKeyName {
			found = true
			break
		}
	}
	// spKeyName has been used before i.e. FixName caused a collision.
	// Add unique postfix: use number of foreign keys in this table so far.
	// However, there is a chance this has already been used,
	// so need to iterate.
	if found {
		id := len(spKeys)
		for {
			c := spKeyName + "_" + strconv.Itoa(id)

			// check wether the spKeyName is used before
			found = false
			for _, spKey := range spKeys {
				if spKey.Name == c {
					found = true
					break
				}
			}
			if !found {
				spKeyName = c
				break
			}
			id++
		}
	}

	return spKeyName, nil
}
```

the only difference with GetSpannerCol and GetSpannerTable is that we don't have direct access to foreignKeys using map, so we need to add one loop to check whether the generated spanner key name is unique.